### PR TITLE
expose space_mt and index_mt on `box` table

### PR DESCRIPTION
**Pull request by aleclarson**
This commit allows userland to extend the space and index metatables
with their own functions or even metamethods. Reducing barriers for
this kind of experimentation is vital for user contribution toward
the improvement of Tarantool's API.

Also, this commit reuses functions between all spaces/indexes.

The metatables are exposed on `box` to make them simple to discover.
The suggested alternatives were `box.internal` and `box.schema`.

In a future commit, we should look into "freezing" the built-in
methods so the user cannot accidentally override them. This is
especially important when updating Tarantool.

### CODE REVIEW

File: src/box/lua/schema.lua [hunk 1]
```diff
@@ -1078,8 +1078,134 @@ end
 
 internal.check_iterator_type = check_iterator_type -- export for net.box
 
-function box.schema.space.bless(space)
+box.space_mt = (function()
```
Comment (Gerold103) [inline src/box/lua/schema.lua]: Please, move it to a box.schema instead of box.

File: src/box/lua/schema.lua [hunk 2]
```diff
@@ -1078,8 +1078,134 @@ end
...
+        if s == nil then
+            box.error(box.error.NO_SUCH_SPACE, space.name)
+        end
+        builtin.space_run_triggers(s, yesno)
+    end
+
+    return space_mt
+end)()
+
+box.index_mt = (function()
```
Comment (Gerold103) [inline src/box/lua/schema.lua]: We do not use in tarantool functions that are called right after creation. Please, initialise box.index_mt with no closures. For example, like this:
```Lua
local index_mt = {}
local function index_mt:len()
-- 'index' here can be accessed as 'self'.
end
local function index_mt:min_ffi(key)
-- ..
end
--..
--..
box.schema.index_mt = { __index = index_mt }
```

File: src/box/lua/schema.lua [hunk 3]
```diff
@@ -1099,8 +1225,6 @@ function box.schema.space.bless(space)
         return tonumber(ret)
     end
     index_mt.__len = index_mt.len -- Lua 5.2 compatibility
-    index_mt.__newindex = function(table, index)
```
Comment (Gerold103) [inline src/box/lua/schema.lua]: Add a test for it, please.

File: src/box/lua/schema.lua [hunk 4]
```diff
@@ -1292,142 +1416,33 @@ function box.schema.space.bless(space)
...
+    index_mt.__pairs = index_mt.pairs -- Lua 5.2 compatibility
+    index_mt.__ipairs = index_mt.pairs -- Lua 5.2 compatibility
+
+    return index_mt
+end)()
+
+function box.schema.space.bless(space)
+    local index_mt = {}
+    index_mt.__index = function(index, key)
+      return index_mt[key] or box.index_mt[key]
```
Comment (Gerold103) [inline src/box/lua/schema.lua]: Why do you need index_mt[key], if all of your extensions must be in box.schema.index_mt ?

**Comment by Gerold103** [comment]
@aleclarson please, check my review, and fix these minor remarks. Thank you for contributing!

File: src/box/lua/schema.lua (hunk 3 above)
Comment (aleclarson) [reply src/box/lua/schema.lua]: Testing what? Adding a new property to `index_mt` and `space_mt`?

Comment (Gerold103) [reply src/box/lua/schema.lua]: Yes. That it does not raise error.

**Comment by kostja** [comment]
Alec, do you still care about this pull request?

**Comment by aleclarson** [comment]
Which directory should I put the test in?

**Comment by aleclarson** [comment]
Tests are passing on my end!

**Comment by aleclarson** [comment]
Thanks for merging this! 👍

**Comment by Mons** [comment]
Please, **don't** create global and single metatable for all spaces!
Currently there is a possibility to ass different features to different spaces
Ex: https://github.com/moonlibs/xqueue

**Comment by aleclarson** [comment]
@Mons I agree. Each space should get its own metatable like so:

```lua
local space_mt = {}
setmetatable(space_mt, {__index = box.schema.space_mt})
setmetatable(space, space_mt)
```

**Comment by aleclarson** [comment]
@Mons You can manually give a space its own metatable like this:

```lua
local shared_mt = getmetatable(space)
local space_mt = setmetatable({}, {__index = shared_mt})
setmetatable(space, space_mt)
```

Why isn't that good enough for your use case?

**Comment by bigbes** [comment]
@aleclarson this won't work, since mt on mt will give `space_mt` as `self`, not `space`. But nothing stands in the way of storing per-space methods in the space object and global methods in the `__index` field of metatable.

**Comment by Mons** [comment]
Why to break backward compatibility, if you introduce new functionality?
Just add metatable to space metatables (as do me)
```lua
local superspace_mt = {}

for id,space in pairs(box.space) do
	if type(id) == 'string' then
		setmetatable(getmetatable(space),{ __index = superspace_mt })
	end
end
```

**Comment by aleclarson** [comment]
Cool, nice job on the changes!

**Pull request closed without merging**